pkg/kubelet/certificate/certificate_manager.go

```
@@ -0,0 +1,244 @@
+// the background it communicates with the API server to get new certificates
+// for certificates about to expire.
+type Manager interface {
+	// Start the API server status sync loop.
+	Start()
+	// GetCertificate gets the current certificate from the certificate
+	// manager. This function matches the signature required by
+	// tls.Config.GetCertificate so it can be passed as TLS configuration. A
+	// TLS server will automatically call back here to get the correct
+	// certificate when establishing each new connection.
+	GetCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error)
+}
+
+// Store is responsible for getting and updating the current certificate.
+// Depending on the concrete implementation, the backing store for this
+// behavior may vary.
+type Store interface {
+	Current() (*tls.Certificate, error)
+	// Accepts the PEM data for the cert/key pair and makes the new cert/key
+	// pair the 'current' pair, that will be returned by future calls to
+	// Current().
+	Update(cert, key []byte) (*tls.Certificate, error)
+}
+
+type manager struct {
+	certSigningRequestClient clientcertificates.CertificateSigningRequestInterface
+	template                 *x509.CertificateRequest
+	usages                   []certificates.KeyUsage
+	certStore                Store
+	certAccessLock           sync.RWMutex
+	cert                     *tls.Certificate
+	shouldRotatePercent      int32
+}
+
+// NewManager returns a new certificate manager. A certificate manager is
+// responsible for being the authoritative source of certificates in the
+// Kubelet and handling updates due to rotation.
+func NewManager(
+	certSigningRequestClient clientcertificates.CertificateSigningRequestInterface,
+	template *x509.CertificateRequest,
+	usages []certificates.KeyUsage,
+	certificateStore Store,
+	certRotationPercent int32) (Manager, error) {
+
+	cert, err := certificateStore.Current()
+	if err != nil {
+		return nil, err
```

Should we instead return an error from GetCertificates() until the store has synced?